File: decentralized-api/broker/openai.go
package broker

// Define the structs that match the JSON structure
type Response struct {
	ID                string   `json:"id"`
	Object            string   `json:"object"`
	Created           int64    `json:"created"`
	Model             string   `json:"model"`
	SystemFingerprint string   `json:"system_fingerprint"`
	Choices           []Choice `json:"choices"`
	Usage             Usage    `json:"usage"`
}

type Choice struct {
	Index    int     `json:"index"`
	Message  Message `json:"message"`
	Logprobs struct {
		Content []Logprob `json:"content"`
	} `json:"logprobs"`
	FinishReason string `json:"finish_reason"`
	StopReason   string `json:"stop_reason"`
}

type Message struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type Logprob struct {
	Token       string  `json:"token"`
	Logprob     float64 `json:"logprob"`
	Bytes       []int   `json:"bytes"`
	TopLogprobs []struct {
		Token   string  `json:"token"`
		Logprob float64 `json:"logprob"`
		Bytes   []int   `json:"bytes"`
	} `json:"top_logprobs"`
}

type Usage struct {
	PromptTokens     uint64 `json:"prompt_tokens"`
	CompletionTokens uint64 `json:"completion_tokens"`
	TotalTokens      uint64 `json:"total_tokens"`
}

File: testermint/src/main/kotlin/data/openai.kt
package com.productscience.data

data class RequestModel(
    val model: String,
    val messages: List<Message>,
    val frequencyPenalty: Int,
    val logitBias: Any?,
    val logprobs: Boolean,
    val topLogprobs: Int,
    val maxTokens: Int,
    val n: Int,
    val presencePenalty: Int,
    val responseFormat: ResponseFormat,
    val seed: Int,
    val serviceTier: Any?,
    val stop: String,
    val stream: Boolean,
    val streamOptions: Any?,
    val temperature: Int,
    val topP: Int,
    val tools: List<Tool>,
    val toolChoice: String,
    val parallelToolCalls: Boolean,
    val user: String,
    val functionCall: String,
    val functions: List<Function>
)

data class Message(
    val content: String,
    val role: String,
    val name: String
)

data class ResponseFormat(
    val type: String
)

data class Tool(
    val type: String,
    val function: FunctionDetails
)

data class FunctionDetails(
    val name: String,
    val description: String,
    val parameters: Map<String, Any>,
    val strict: Boolean
)

data class Function(
    val name: String,
    val description: String,
    val parameters: Map<String, Any>
)

// Response

data class OpenAIResponse(
    val choices: List<Choice>,
    val created: Long,
    val id: String,
    val model: String,
    val `object`: String,
    val usage: Usage
)

data class Choice(
    val finishReason: String,
    val index: Int,
    val logprobs: Logprobs?,
    val message: ResponseMessage,
    val stopReason: Any?
)

data class Logprobs(
    val content: List<Content>
)

data class Content(
    val bytes: List<Int>,
    val logprob: Double,
    val token: String,
    val topLogprobs: List<TopLogprob>
)

data class TopLogprob(
    val bytes: List<Int>,
    val logprob: Double,
    val token: String
)

data class ResponseMessage(
    val content: String,
    val role: String,
    val toolCalls: List<Any>
)

data class Usage(
    val completionTokens: Int,
    val promptTokens: Int,
    val totalTokens: Int
)
